Add quick tip on neural keyword extraction

kbuschme opened the pull request.

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/README.md:
```diff
@@ -0,0 +1,7 @@
+# Neural Keyword Extraction
+
+…
```
Perhaps a small description here?

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "!python -m spacy download en # Download English model"
+      ],
+      "execution_count": 23,
+      "outputs": [
```
Perhaps for cleanliness we can clear the output of this cell. 
It's also worth it to add a `-q` flag to the pip install statement to generally suppress output

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "\n",
+        "Strictly speaking [KeyBERT](https://github.com/MaartenGr/KeyBERT) is not an end-to-end neural keyword extraction model. \n",
+        "\n",
+        "Nontheless the **underlying idea** is as clever as it is simple: It compares embeddings of words with embeddings of texts and selects the set of keywords which are most similar to the entire text.\n",
```
small typo: nontheless > nonetheless

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "Nontheless the **underlying idea** is as clever as it is simple: It compares embeddings of words with embeddings of texts and selects the set of keywords which are most similar to the entire text.\n",
+        "Both **word embeddings** as well was **text embeddings** are generated using **state-of-the-art neural models**, which have lead to tremendous performance improvements in other tasks.\n",
+        "\n",
+        "The benefit of this approach is that the used models are trained in an unsupervised manner and thus **do not require an annotated dataset** if keywords. In addition, they are available in a number of **non-English languages** as well.\n",
```
small typo: `if keywords` > `of keywords`

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "\n",
+        "🤗 Hugging Face Model Hub is *the* prime address when it comes to discovering state-of-the-art models that are easy to use. However, with respect to neural keyword (or keyphrase) extraction there are only two models available as of this writing.\n",
+        "\n",
+        "Both models were created by [Ankur Singh](https://huggingface.co/ankur310794) and use a **BART-based sequence-to-sequence architecture**. Unfortunately is not a lot of details available about the training specific \n",
```
Bit of a strange sentence: `Unfortunately is not a lot of details available about the training specific` > `Unfortunately there are not a lot of details available about the training specifics`

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "\n",
+        "[One model](https://huggingface.co/ankur310794/bart-base-keyphrase-generation-kpTimes) was trained using the [KPTimes dataset](https://aclanthology.org/W19-8617/), a large dataset consisting of **English news articles** and hand-annotated keywords.\n",
+        "\n",
+        "[The other](https://huggingface.co/ankur310794/bart-base-keyphrase-generation-openkp) was trained using the [OpenKP dataset](https://github.com/microsoft/OpenKP), which contains a large number of **English web documents** and up to three most relevant keywords. From this This restriction holds true for the model as well: it will return at most three keywords."
```
I think there is a `From this` too much in the last sentence

TDehaene replied:
bullet point

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "        \"Los Angeles. “But we don’t want to change the way we do things \"\n",
+        "        \"because of the hype.”\"\n",
+        "    ),\n",
+        "    # The text within this notebook:\n",
```
Cleverly done 😄

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "    compare_keyword_extraction_algorithms(text, selection=selected_algorithms)"
+      ],
+      "execution_count": 66,
+      "outputs": [
```
it's a bit strange how the BART-based models are supposed to return the top-3 keywords, but seem to return 2, 2, 6, 2 and 2 keywords respectively, any idea what might cause this?

TDehaene replied:
Awesome, thanks 👍

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "id": "9wfoRIn1zPQD"
+      },
+      "source": [
+        "To-do: \n",
```
Are these items still to do, or are they remnants from the WIP?

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+              "    <tr>\n",
+              "      <th>0</th>\n",
+              "      <td>methods unable</td>\n",
+              "      <td>Search Engines</td>\n",
```
Funny how the BART-based method seems to dream up certain keywords, saw this popping up quite some times when playing around with it. Would definitely mention this somewhere in the conclusion or in the intro on the BART-based methods

kbuschme replied:
I think this this depends on the dataset actually. One dataset only contains three keywords, the other contains more. Good remark though, I'll try to clarify this above!

kbuschme replied:
Do you mean that some of the words don't actually appear in the text? This is one of the potential benefits of the neural approaches: keyword generation or abstractive extraction of you will.

TDehaene replied:
That's also definitely an interesting way of looking at it, it would seem that the knife cuts both-ways:
- a benefit is indeed that it can create abstractive keywords
- a disadvantage is that the abstractive keywords are sometimes not perfectly related to the text in question (like in this case)

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "\n",
+        "We further presented two approaches that are readily available: **KeyBERT** uses the **similarity** between word and text **embeddings** to find keywords which best describe a text and an **end-to-end neural keyword extraction model** based on BART.\n",
+        "\n",
+        "For now, we recommend to first try `pke` to establish a solid baseline and explore the mentioned extraction methods as an addition.\n",
```
Solid recommendation, would put this statement a bit more to the foreground through an 👉 emoji before it

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "\n",
+        "## Summary\n",
+        "\n",
+        "This notebook gave a brief overview of neural keyword extraction.\n",
```
Would put the two following sentences `If an annotated...` and `We further presented...` as indented bullet points under this sentence for readability

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "id": "j78lHqRwsUeU"
+      },
+      "source": [
+        "In a [pervious quick tip](https://github.com/ml6team/quick-tips/tree/main/nlp/2021_03_18_pke_keyword_extraction) we looked at [`pke`](https://boudinfl.github.io/pke/build/html/index.html) as a replacement for Gensim's [recently removed keyword extraction module](https://github.com/RaRe-Technologies/gensim/releases/tag/4.0.0).\n",
```
For readability, I would chop up the first section in a few subsections and add a little bit more bullets. Hopefully the following image makes this a bit clear:
![image](https://user-images.githubusercontent.com/22496343/124916776-56214500-dff3-11eb-83ad-702dff438838.png)

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "Strictly speaking [KeyBERT](https://github.com/MaartenGr/KeyBERT) is not an end-to-end neural keyword extraction model. \n",
+        "\n",
+        "Nontheless the **underlying idea** is as clever as it is simple: It compares embeddings of words with embeddings of texts and selects the set of keywords which are most similar to the entire text.\n",
+        "Both **word embeddings** as well was **text embeddings** are generated using **state-of-the-art neural models**, which have lead to tremendous performance improvements in other tasks.\n",
```
small typo: `as well was` > `as well as`

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+        "\n",
+        "Both models were created by [Ankur Singh](https://huggingface.co/ankur310794) and use a **BART-based sequence-to-sequence architecture**. Unfortunately is not a lot of details available about the training specific \n",
+        "\n",
+        "[One model](https://huggingface.co/ankur310794/bart-base-keyphrase-generation-kpTimes) was trained using the [KPTimes dataset](https://aclanthology.org/W19-8617/), a large dataset consisting of **English news articles** and hand-annotated keywords.\n",
```
bullet point

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1520 @@
+      "metadata": {
+        "id": "fr1oYx4_tMrp"
+      },
+      "source": [
```
extreme nitpick: unused last cell

TDehaene reviewed (changes requested):
All in all a great tip! Content-wise I don't have any remarks, I think this tip is a great addition to the collection and provides valuable insights 👌 !

Some comments on typos and layout, and perhaps a small side-note that we should mention something on the typical BART-weirdness that can occur.

But again, in general: strong stuff 💪

kbuschme commented:
Thanks for the review, @TDehaene! I added a description to the README and fixed the typos/layout issues. Somehow the file format of the notebook is different now, so everything is marked as outdated. 😞

Any feedback is still very welcome! 😀

ThomasVrancken commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1373 @@
+   "source": [
+    "## KeyBERT\n",
+    "\n",
+    "Strictly speaking [KeyBERT](https://github.com/MaartenGr/KeyBERT) is not an end-to-end neural keyword extraction model. \n",
```
I was a bit surprised to see that you originally say that the NKE methods are based on RNNs/LSTMs and then the first two models are Transformers ones 😛 Would it be worth mentioning that some use transformers/attention mechanism architectures?

ThomasVrancken commented on nlp/2021_07_07_neural_keyword_extraction/README.md:
```diff
@@ -0,0 +1,11 @@
+
+We recommend to open the notebook using Colab for an interactive explainable experience and optimal rendering of the visuals 👇:
+
+[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ml6team/quick-tips/blob/main/nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb)
```
Link didn't work for me

kbuschme replied:
That's expected as the pull request is not yet merged and thus the notebook is not part of the main branch. :)

TDehaene replied:
placeholder link for PR colab:  https://colab.research.google.com/github/ml6team/quick-tips/blob/feature/nlp/neural-keyword-extraction/nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb

ThomasVrancken commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1373 @@
+    "# Compare the keywords extracted by the given algorithms\n",
+    "selected_algorithms = ['KeyBERT', 'BART-based', 'TFIDF','YAKE', 'TextRank', 'TopicRank']\n",
+    "\n",
+    "for text in texts:\n",
```
First thing that occurred to me is why BART gets only a couple of keywords. I saw Thomas D remark about that but couldn't see the explanation in the markdowns (maybe I just missed it though 🙈). Maybe put a line/comment about that?

kbuschme replied:
Do you mean for both of the BART-based extraction methods? Above I mention the that for one of the two models the maximum number of keywords is 3 because the dataset contains a maximum of 3 keywords per training document.

But since you both were confused I'll mention it here once more.

ThomasVrancken reviewed (commented):
Very nice one! The content is useful and interesting (had to implement the pke methods manually a while back, cool to see that there are more deep learning approaches now). And the notebook is clear and thorough. I really liked that you show the different alternatives, explain their differences and compare them. No big feedback from my part, I think Thomas D already covered it. Looking forward to see it go live 😄

TDehaene commented on nlp/2021_07_07_neural_keyword_extraction/neural_keyword_extraction.ipynb:
```diff
@@ -0,0 +1,1373 @@
+    "In a [pervious quick tip](https://github.com/ml6team/quick-tips/tree/main/nlp/2021_03_18_pke_keyword_extraction) we looked at [**pke**](https://boudinfl.github.io/pke/build/html/index.html) as a replacement for Gensim's [recently removed keyword extraction module](https://github.com/RaRe-Technologies/gensim/releases/tag/4.0.0).\n",
+    "pke comes with batteries included: \n",
+    "\n",
+    "* It has **preprocessing build in**, \n",
```
small typo: build in > built-in

kbuschme commented:
Thanks for the additional reviews!

As advised, I added a paragraph highlighting some aspects of the results and mentioned the potentially abstractive keywords in the summary as well.

Feel free to have another look. Any feedback is still very welcome.

TDehaene commented:
LGTM :-), you'll have a merge conflict due to other tip being merged as well
